specs/exp-proof.md

```diff
@@ -0,0 +1,67 @@
+# Exponentiation Proof
+
+The exponentiation proof validates entries in the exponentiation circuit, which includes the exponentiation table. Each `EXP` opcode performed in a block's transactions is mapped to an exponentiation trace, more specifically these are the intermediate multiplication steps produced while exponentiation by squaring.
+
+For instance, `3 ^ 13 == 1594323 (mod 2^256)` is mapped to the following steps:
+```
+3      * 3   = 9
+9      * 3   = 27
+27     * 27  = 729
+729    * 729 = 531441
+531441 * 3   = 1594323
+```
+
+Moving on, we refer each of the above multiplication steps as:
+```
+a * b + c = d
+```
+while validating the equation using a `MulAddWordsGadget` where `c == 0`.
+
+For a detailed algorithm, please refer the [design document](https://hackmd.io/@rohitnarurkar/BJhpYGiCc).
+
+## Circuit Layout
+
+The exponentiation circuit contains columns from the [exponentiation table](./tables.md#exponentiation-table) in addition to the following columns:
+1. `q_step`: A selector to indicate whether or not the current row represents a step in the exponentiation trace.
+2. `is_pad`: A boolean-valued advice column to indicate whether or not the step is reserved for padding or not.
+3. `is_odd`: a boolean-valued advice column to indicate whether or not the exponent at this step is odd or even.
+4. `quotient`: An RLC field representing the quotient in `2 * q + is_odd == exponent`.
+
+We even use a few columns to perform multiplication (modulo `2^256`) of the intermediate steps in the exponentiation trace.
+
+## Circuit Constraints
+
+- For every step except the last step, validate that:
+    - `base` MUST be the same across subsequent steps.
+    - Multiplication result `d` from the next row MUST be equal to the first multiplicand `a` in the current row. For instance, if we consider `row_0` (`531441 * 3 = 1594323`) and `row_1` (`729 * 729 = 531441`) from the above example, `d_1` is `531441` and `a_0` is also `531441`.
+    - `identifier` MUST be the same across subsequent steps.
+
+- For every step, validate that:
+    - `is_last` MUST be boolean.
```

Comment:
What does `is_last` mean?
If I have two instances to prove:

1. 3^13
2. 5^17

Do we set `is_last =1` at the end each of the instances 3^13 and 5^17? Or do we do so only at the end of the last instance 5^17?

Reply:
In the case of the exponentiation circuit, `is_last` is set as true on the row representing the last entry in the exponentiation trace. Which should always be `a * a == a^2` (for `a ^ b == d`).

There will be exactly one row per trace where `is_last == 1`. So for `n` traces in the entire circuit, there will be `n` rows where `is_last == 1`.

Example, if we have two exponentiation operations in a single block, namely:
1. `3 * 3 == 27`
2. `4 * 3 == 64`

Then we would have (a simplified version):
| is_first | is_last | base | exponent | exponentiation |
|--------|--------|------|----------|----------------|
| 1           | 0          | 3       | 3              | 27                      |
| 0          | 1           | 3       | 2              | 9                        |
| 1           | 0          | 4       | 3              | 64                     |
| 0          | 1           | 4       | 2              | 16                      |

So it is at the end of each instance. I will add this to the design document to make it clearer :)

Reply:
I saw the design doc and it makes sense now.